src/dbus/mod.rs

```diff
@@ -0,0 +1,94 @@
+//! D-Bus server interface actor.
+
+use crate::update_agent::{LastRefresh, UpdateAgent};
+use actix::prelude::*;
+use actix::Addr;
+use core::convert::TryFrom;
+use futures::executor;
+use futures::prelude::*;
+use log::trace;
+use zbus::{dbus_interface, fdo};
+use zvariant::ObjectPath;
+
+/// Experimental interface for testing.
+struct Experimental {
+    agent_addr: Addr<UpdateAgent>,
+}
+
+#[dbus_interface(name = "org.coreos.zincati1.Experimental")]
+impl Experimental {
+    /// Just a test method.
+    fn moo(&self, talkative: bool) -> String {
+        if talkative {
+            String::from("Moooo mooo moooo!")
+        } else {
+            String::from("moo.")
+        }
+    }
+
+    /// Get update_agent actor's last refresh time.
+    fn last_refresh_time(&self) -> Result<i64, zbus::fdo::Error> {
+        let msg = LastRefresh {};
+        let refresh_time_fut = self
+            .agent_addr
+            .send(msg)
+            .unwrap_or_else(|e| Err(e.into()))
+            .map_err(|e| {
+                let err_msg = format!("failed to get last refresh time from agent actor: {}", e);
+                log::error!("{}", err_msg);
+                zbus::fdo::Error::Failed(err_msg)
+            });
+
+        executor::block_on(refresh_time_fut)
+    }
+}
+
+pub struct DBusService {
+    agent_addr: Addr<UpdateAgent>,
+}
+
+impl DBusService {
+    /// Create new DBusService
+    fn new(agent_addr: Addr<UpdateAgent>) -> DBusService {
+        DBusService { agent_addr }
+    }
+
+    /// Start the threadpool for DBusService actor.
+    pub(crate) fn start(threads: usize, agent_addr: Addr<UpdateAgent>) -> Addr<Self> {
+        SyncArbiter::start(threads, move || DBusService::new(agent_addr.clone()))
+    }
+}
+
+impl Actor for DBusService {
+    type Context = SyncContext<Self>;
+
+    fn started(&mut self, _ctx: &mut Self::Context) {
+        trace!("D-Bus service actor started");
+
+        let connection = zbus::Connection::new_system().unwrap();
```

Comment:
Isn't there a way to more properly report errors from actor startup, like a way to return a `Result`?  I am not finding it offhand; looks like `MetricsService` is calling `expect()` (which is an unwrap-wrapper).

At least one thing I'd do in this situation is factor out a helper function that returns a `Result`, and then centralize something like a `log::error!(e); return;` call from that.